### PR TITLE
Heat transfer assembler

### PR DESCRIPTION
**Pull request by shahabgol**
This PR takes care of refactoring assemblers in heat transfer solver.
Three separate assemblers, namely core assembler, BDF and Robin BC assemblers are defined.

### CODE REVIEW

File: include/solvers/heat_transfer.h [hunk 1]
```diff
@@ -19,7 +19,8 @@
  * Equation solved:
  * rho * Cp * (dT/dt + u.gradT) = k div(gradT) + nu/rho * (gradu : gradu)
  *
- * Author: Bruno Blais and Jeanne Joachim, Polytechnique Montreal, 2020-
+ * Author: Bruno Blais, Jeanne Joachim and Shahab Golshan, Polytechnique
+ Montreal, 2020-
```
Comment (blaisb) [inline include/solvers/heat_transfer.h:23]:
```suggestion
```
Let's remove the authors now. I think it's a better approach.

File: include/solvers/heat_transfer_assemblers.h [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+ * ---------------------------------------------------------------------
+
+ *
+ * Implementation of heat transfer as an auxiliary physics.
+ * This heat equation is weakly coupled to the velocity field.
+ * Equation solved:
+ * rho * Cp * (dT/dt + u.gradT) = k div(gradT) + nu/rho * (gradu : gradu)
+ *
+ * Author: Bruno Blais, Jeanne Joachim and Shahab Golshan, Polytechnique
+ Montreal, 2020-
```
Comment (blaisb) [inline include/solvers/heat_transfer_assemblers.h:23]:
```suggestion
```

File: include/solvers/heat_transfer_assemblers.h [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+template <int dim>
+class HeatTransferAssemblerBase
+{
+public:
+  /**
+   * @brief assemble_matrix Interface for the call to matrix assembly
+   * @param scratch_data Scratch data containing the heat transfer information.
+   * It is important to note that the scratch data has to have been re-inited
+   * before calling for matrix assembly.
+   * @param copy_data Destination where the local_rhs and loc
```
Comment (blaisb) [inline include/solvers/heat_transfer_assemblers.h:52]: Incomplete comment here

File: include/solvers/heat_transfer_assemblers.h [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  virtual void
+  assemble_matrix(HeatTransferScratchData<dim> &scratch_data,
+                  StabilizedMethodsCopyData &   copy_data) = 0;
+
+  /**
+   * @brief assemble_matrix Interface for the call to rhs
+   * @param scratch_data Scratch data containing the heat transfer information.
+   * It is important to note that the scratch data has to have been re-inited
+   * before calling for matrix assembly.
+   * @param copy_data Destination where the local_rhs and loc
```
Comment (blaisb) [inline include/solvers/heat_transfer_assemblers.h:64]:
```suggestion
   * @param copy_data Destination where the element for the local_rhs and local_matrix are copied to
```

File: include/solvers/heat_transfer_assemblers.h [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+   * @param copy_data (see base class)
+   */
+  virtual void
+  assemble_rhs(HeatTransferScratchData<dim> &scratch_data,
+               StabilizedMethodsCopyData &   copy_data) override;
+
+
+  std::shared_ptr<SimulationControl> simulation_control;
+  Parameters::PhysicalProperties     physical_properties;
+  const SimulationParameters<dim> &  simulation_parameters;
```
Comment (blaisb) [inline include/solvers/heat_transfer_assemblers.h:119]: What are simulation parameters and why are they needed? If you have simulation control and physical properties?

File: include/solvers/heat_transfer_scratch_data.h [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+
+#include <deal.II/numerics/vector_tools.h>
+
+using namespace dealii;
+
+
+/**
+ * @brief TracerScratchData class
+ * stores the information required by the assembly procedure
+ * for a Tracer advection-diffusion equation. Consequently, this class
```
Comment (blaisb) [inline include/solvers/heat_transfer_scratch_data.h:50]:
```suggestion
 * for a heat transfer equation. Consequently, this class
```

File: include/solvers/heat_transfer_scratch_data.h [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+ *calculates the tracer (values, gradients, laplacians) and the shape function
+ * (values, gradients, laplacians) at all the gauss points for all degrees
+ * of freedom and stores it into arrays. Additionnaly, the use can request
+ * that this class gathers additional fields for physics which are coupled
+ * to the Tracer equation, such as the velocity which is required. This class
```
Comment (blaisb) [inline include/solvers/heat_transfer_scratch_data.h:55]:
```suggestion
 * to the heat equation, such as the velocity which is required. This class
```

File: source/solvers/heat_transfer.cc [hunk 8]
```diff
@@ -23,589 +27,237 @@
...
+  assemble_system_rhs();
 }
 
 
 template <int dim>
 void
-HeatTransfer<dim>::assemble_rhs(
-  const Parameters::SimulationControl::TimeSteppingMethod time_stepping_method)
+HeatTransfer<dim>::assemble_rhs(const Parameters::SimulationControl::
+                                  TimeSteppingMethod /*time_stepping_method*/)
```
Comment (blaisb) [inline source/solvers/heat_transfer.cc:41]: With the new merge of the KINSOL solver this interface should be deprecated now. I think you can remove this alltogether and just use the assemble_system_rhs

File: source/solvers/heat_transfer.cc [hunk 9]
```diff
@@ -23,589 +27,237 @@
...
-    bdf_coefs = bdf_coefficients(3, time_steps_vector);
-
-  if (time_stepping_method ==
-        Parameters::SimulationControl::TimeSteppingMethod::sdirk22_1 ||
-      time_stepping_method ==
-        Parameters::SimulationControl::TimeSteppingMethod::sdirk33_1)
+  this->assemblers.clear();
+
+  // Robin boundary condition
+  this->assemblers.push_back(std::make_shared<HeatTransferAssemblerRBC<dim>>(
```
Comment (blaisb) [inline source/solvers/heat_transfer.cc:53]: I would give this class a more explicit name. HeatTransferAssemblerRobinBC or HeatTransferAssemblerConvectionBC for example.

File: source/solvers/heat_transfer_assemblers.cc [hunk 10]
```diff
@@ -0,0 +1,607 @@
...
+
+          thermal_conductivity = calculate_point_property(
+            scratch_data.phase_values[q],
+            physical_properties.fluids[0].thermal_conductivity,
+            physical_properties.fluids[1].thermal_conductivity);
+
+          // Useful definitions
+          rho_cp = density * specific_heat;
+          alpha  = thermal_conductivity / rho_cp;
+        }
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:71]: I guess this is why you passed the simulation parameters? I'm not sure this is the best approach. Maybe then pass the multiphysics parameters, but just pass the minimum needed to the class. I can understand that you want to keep these properties in a  if for now and that makes sense.

File: source/solvers/heat_transfer_assemblers.cc [hunk 11]
```diff
@@ -0,0 +1,607 @@
...
+            {
+              const Tensor<1, dim> grad_phi_T_j = scratch_data.grad_phi_T[q][j];
+
+
+              // Weak form for : - k * laplacian T + rho * cp *
+              //                  u * gradT - f -
+              //                  tau:grad(u) =0
+              // Hypothesis : incompressible newtonian fluid
+              // so tau:grad(u) =
+              // mu*(grad(u)+transpose(grad(u)).transpose(grad(u))
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:112]: Viscous dissipation should not be in the core assembler (and it is not right now no?)

File: source/solvers/heat_transfer_assemblers.cc [hunk 12]
```diff
@@ -0,0 +1,607 @@
...
+                                      scratch_data.phi_face_T[j] * h * JxW;
+                                  }
+                              }
+                          }
+                      }
+                    }
+                }
+            }
+        }
+    } // end loop for Robin condition
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:528]:
```suggestion
    } 
```

File: source/solvers/heat_transfer_assemblers.cc [hunk 13]
```diff
@@ -0,0 +1,607 @@
...
+                                   T_inf) *
+                                  JxW;
+                              }
+                          }
+                      }
+                    }
+                }
+            }
+        }
+    } // end loop for Robin condition
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:603]:
```suggestion
    } 
```

File: source/solvers/heat_transfer_assemblers.cc [hunk 14]
```diff
@@ -0,0 +1,607 @@
...
+            JxW;
+
+          if (this->simulation_parameters.multiphysics.viscous_dissipation)
+            {
+              local_rhs(i) -= (-dynamic_viscosity * phi_T_i *
+                               scalar_product(velocity_gradient +
+                                                transpose(velocity_gradient),
+                                              transpose(velocity_gradient))) *
+                              JxW;
+            }
```
Comment (blaisb) [inline source/solvers/heat_transfer_assemblers.cc:244]: I would put this into another assembler that just assembles an RHS instead of putting an if there.

**Review by blaisb — COMMENTED** [review]
Looks very good.
Close to being ready for a merge.
1- Just remove the whole simulation parameters given to to the assembler and just give them the multiphysics parameters
2- Put viscous dissipation in another separate assembler. It will be cleaner this way. I would like to avoid having confusing ifs in the assembly.

Then this should be ready for merge :)

**Comment by shahabgol** [comment]
All comments are applied. Viscous dissipation is defined in a separate assembler, and multiphysics parameters is used instead of simulation parameters.

**Comment by blaisb** [comment]
@shahabgol  There are two conflicts to handle, please merge or rebase

**Review by blaisb — APPROVED** [review]
Looks good! Just need to fix conflicts before merge can be done :)

**Comment by blaisb** [comment]
There still seems to be some conflicts to fix, we can talk about it this week if you need help :)!

**Comment by shahabgol** [comment]
> There still seems to be some conflicts to fix, we can talk about it this week if you need help :)!

Fixed. :)

**Review by blaisb — APPROVED** [review]

**Pull request merged**